### ISSUE TITLE
Test-based shader warmup

### ISSUE THREAD

**Issue by liyuqian** [flutter/flutter#53609]
This tracks our medium-term solution of #32170 that's based on something similar to driver tests.

This approach would be completely device/hardware-independent.

Also, it will work well on newer backends such as Vulkan and Metal, whereas our short term solution "SkSL-based shader warmup #53607" would have issues.

Currently this is blocked on https://github.com/flutter/flutter/issues/38329

**Comment by kf6gpe** [comment]
cc @kf6gpe

**Comment by zanderso** [comment]
@liyuqian are you working on this? Should this issue have a priority and milestone? Thanks!

**Comment by liyuqian** [comment]
@zanderso : this is currently blocked by https://github.com/flutter/flutter/issues/38329 . I'll self-assign and add a milestone once that's closed.

**Comment by acoutts** [comment]
Looking forward to this one!

**Comment by dannyvalentesonos** [comment]
@liyuqian I'm having trouble understanding what this solution would look like in practical terms for developers. Will we still need to run some tests and package some output into the actual app? If so, will this solution work with Add2App?
Thanks!

**Comment by liyuqian** [comment]
@dannyvalentesonos : You'll write some tests like https://github.com/flutter/flutter/pull/62171/files#diff-e05d298bb504bbbed0b6f263fd72e954R17-R27. There's no need to run the test locally or generate something from it. We'll directly package the test code into the actual app, run the test code in the background with some special GPU context that only compiles shaders. It should work with Add2App.

**Comment by dannyvalentesonos** [comment]
That's great. Thanks @liyuqian . Do you have any sense of when this might be ready? I'm not looking for any accurate date, but wondering if this is possibly something in 2020 or 2021. Thanks.

**Comment by liyuqian** [comment]
It's probably 2021.

**Comment by xinoxapps** [comment]
Any change this can have a boost with the available resources of Google. 

This is a big deal breaker as animations become more popular in the onboarding screens of the app.

**Comment by kf6gpe** [comment]
Escalating to meet priority of umbrella bug and customer feedback.

**Comment by Superkramar** [comment]
Folks, how this issue can be P3 priority???  We won't be able to go to production with junk animation. And a have read a lot of comments with same problem. I just want to say that you should rethink about priority. It is obviously P1.

**Comment by meowofficial** [comment]
@Superkramar Someone from Flutter team said that this is highest possible priority for issues that don't break build and aren't faced by priority customers.

**Comment by Superkramar** [comment]
I think it will be exception from rules. Or they need to explain it for my Project manager =)))

**Comment by meowofficial** [comment]
For me it's the same as issues that break build, because it makes the build unusable for ios.

**Comment by theniceboy** [comment]
This should be the number one priority for the Flutter team.
![Screen Shot 2020-11-11 at 1 07 51 AM](https://user-images.githubusercontent.com/8187501/98706879-56146b00-23ba-11eb-96a6-52a2050a84bb.png)

**Comment by kf6gpe** [comment]
Friends.
I'd like to remind everyone that the purpose of the issue database is for constructive criticism and collaboration around the issues posted. Some of this conversation isn't constructive, and doesn't further the work being done by the team.

I'll also remind everyone of the priorities as laid out on our wiki at https://github.com/flutter/flutter/wiki/Issue-hygiene#priorities. P3 is for issues that are high priority currently being worked on. We hear your interest in seeing this resolved, and are _actively working_ on doing so.

**Comment by liyuqian** [comment]
As mentioned in https://github.com/flutter/flutter/issues/32170#issuecomment-724827948, solving shader compilation jank is among the highest priority, and we're actively working on it. Here's a concrete work in progress in case someone missed it: [supporting Metal binary shader cache is estimated to be done by December][1].

[1]: https://github.com/flutter/flutter/issues/60267#issuecomment-724764147

**Comment by esDotDev** [comment]
A recent poll on twitter shows around 60-70% of Flutter developers run no tests at all. Is this really the ideal approach, to force all this additional code to be written?

For my specific use case, I have 3 views in my entire application that I want to cache, everything else runs perfectly. So, to open these 3 views, and cache the shaders, would take me about 10 seconds and 5 taps with my finger. To simulate navigating my app to all these different views, you're talking about days or more, and a completely new learning curve of testing, simply to capture a json file?

**Comment by liyuqian** [comment]
Developer's experience is very important to us so we're also aiming for an ultimate solution that require neither automated testing, nor manual testing. We'd like to make something magical that just works with 0 extra effort.

It may be nontrivial so we'll also take your concerns into considerations before that ideal solution is ready. For example, maybe we could have tools that record your manual tests into automated tests. Or maybe we can create another recording format that sits between the automated tests and the low-level SkSL presentations. (Or maybe we get lucky and find an ideal solution that can be implemented quickly.)

As you've seen, there are many different approaches to explore and many concerns to consider, so it may take us some time to design and implement the details. We're actively working on it.

**Comment by esDotDev** [comment]
That sound really cool (translating manual user actions into an automated test script). Don't tease me like that :p

**Comment by britannio** [comment]
> That sound really cool (translating manual user actions into an automated test script). Don't tease me like that :p

https://www.repeato.app/ already does it in a way.

**Comment by evetechsolutions** [comment]
Is there any update on this issue as of March 2021? Thank you

**Comment by Hixie** [comment]
I don't really understand what was being proposed here. Does anyone know what @liyuqian had in mind here?

**Comment by zanderso** [comment]
I believe the idea here was to drive the technology from https://github.com/flutter/flutter/issues/32170 with a set of automated tests run as part of `flutter test` rather than through manual interaction as part of `flutter run`.

**Comment by esDotDev** [comment]
But with a very well liked comment saying that maybe could still use `flutter run` to generate those test files, because developers don't actually want to write test code for this sort of thing. 

A fairly reasonable dev experience looks like:
1. Open the app in profile mode
2. Do a 1m smoke test, open every screen in the app, trigger all the transitions you want
3. Some sort of platform agnostic automated test code is created, flutter can use to pre-warm the app in any environment

Basically make the developer the integration test_runner, the system watches him and records his actions, and there's your test.

**Comment by Hixie** [comment]
@zanderso do we know what's needed to make that happen? Is it basically "run flutter_tester with all the backends enabled"?

@esDotDev that's already possible right?

**Comment by zanderso** [comment]
It sounds like @esDotDev is asking for [SkSL warmup](https://flutter.dev/docs/perf/rendering/shader#how-to-use-sksl-warmup) plus the ability to record a series of actions that can then be replayed---as opposed to writing something like that manually.

@Hixie That's a good question. It looks like a lot of progress was made on flutter.dev/go/flutter-nonhost. As noted above, there are in-tree examples [here](https://github.com/flutter/flutter/pull/62171/files#diff-e05d298bb504bbbed0b6f263fd72e954R17-R27), so making a public API and giving it good documentation might be all that is left to do, but progress on https://github.com/flutter/flutter/issues/38329 appears to have stalled out.

**Comment by esDotDev** [comment]
> @esDotDev that's already possible right?

I really have no idea. tbh I don't really know how to write tests, how the framework works, or any of that. And to super honest, I have like 0 passion to learn any of it, my passing is building UI's.

So I was thinking more along the lines of "Press this button", I run my app, smoke test it, and everything else is just taken care of somehow by the framework.

**Comment by Hixie** [comment]
@esDotDev  Ah ok. That's not what this bug is about. This bug is about being able to create automated tests to collect shaders.

**Comment by esDotDev** [comment]
Just recap my understanding:

* This bug is about the ability to write test code, that can be used to warm up shaders in a device independent way (I'm assuming this is some test code, that is run on device, just in time, to warm up shaders so the app then runs smoothly).
* I'm simply saying, the _write test code_ portion of this is problematic, maybe there some way we can automate it via recording a developer actions

**Comment by Hixie** [comment]
> * (I'm assuming this is some test code, that is run on device during the splash screen, just in time, to warm up shaders so the app then runs smoothly).

Test code runs on the developer's computer (or in CI). The output would be something that you can package with the app so that the app can warm up shaders on launch, like you can do now with the more manual warm-up.

**Comment by Hixie** [comment]
Being able to generate a test that replays a given input would be interesting, can you file a bug suggesting that? (It could be useful outside of shader warmup too.)

**Comment by esDotDev** [comment]
Done: https://github.com/flutter/flutter/issues/78869 
Cheers

**Comment by zanderso** [comment]
SkSL caching and warm-up is no longer the approach the team is pursuing to solve early-onset jank.
